crates/runtime/src/instance/allocator.rs

```diff
@@ -0,0 +1,633 @@
+        if start
+            .checked_add(init.data.len())
+            .map_or(true, |end| end > memory.current_length)
+        {
+            return Err(InstantiationError::Trap(Trap::wasm(
+                ir::TrapCode::HeapOutOfBounds,
+            )));
+        }
+
+        unsafe {
+            let mem_slice = get_memory_slice(init, instance);
+            let end = start + init.data.len();
+            let to_init = &mut mem_slice[start..end];
+            to_init.copy_from_slice(&init.data);
+        }
+    }
+
+    Ok(())
+}
+
+fn check_init_bounds(instance: &Instance) -> Result<(), InstantiationError> {
+    check_table_init_bounds(instance)?;
+
+    match &instance.module.memory_initialization {
+        Some(MemoryInitialization::Paged { .. }) | None => {
+            // Bounds were checked at compile-time
+        }
+        Some(MemoryInitialization::OutOfBounds) => {
+            return Err(InstantiationError::Link(LinkError(
+                "memory out of bounds: data segment does not fit".into(),
+            )));
+        }
+        Some(MemoryInitialization::Segmented(initializers)) => {
+            check_memory_init_bounds(instance, initializers)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn initialize_instance(
+    instance: &Instance,
+    is_bulk_memory: bool,
+) -> Result<(), InstantiationError> {
+    // If bulk memory is not enabled, bounds check the data and element segments before
+    // making any changes. With bulk memory enabled, initializers are processed
+    // in-order and side effects are observed up to the point of an out-of-bounds
+    // initializer, so the early checking is not desired.
+    if !is_bulk_memory {
+        check_init_bounds(instance)?;
+    }
+
+    // Initialize the tables
+    initialize_tables(instance)?;
+
+    // Initialize the memories
+    match &instance.module.memory_initialization {
+        Some(MemoryInitialization::Paged { page_size, map }) => {
+            for (index, pages) in map {
+                let memory = instance.memory(index);
+
+                for (page_index, page) in pages.iter().enumerate() {
+                    if let Some(data) = page {
+                        // Bounds checking should have occurred when the module was compiled
+                        // The data should always be page sized
+                        assert!((page_index * page_size) < memory.current_length);
+                        assert_eq!(data.len(), *page_size);
+
+                        unsafe {
+                            ptr::copy_nonoverlapping(
+                                data.as_ptr(),
+                                memory.base.add(page_index * page_size),
+                                data.len(),
+                            );
+                        }
+                    }
+                }
+            }
+        }
+        Some(MemoryInitialization::OutOfBounds) => {
+            return Err(InstantiationError::Trap(Trap::wasm(
```

Comment:
This condition (i.e. `MemoryInitialization::OutOfBounds`) only occurs when *all* the memories being initialized are defined (i.e. no imported memories).

As such, there can't be any visible side-effects to the initialization because the module never successfully instantiates and no linear memory is being written to that existed prior to instantiation.

Reply:
Ah I forgot yeah that this only happens when all memories are being defined. That being said I think weird spec corner cases will require us to still perform in-order initialization.

According to https://webassembly.github.io/bulk-memory-operations/core/exec/modules.html#instantiation the 10th step is table initialization and the 11th step is data segment initialization. That means that we could initialize imported tables with functions this module defines (which close over defined memories). I believe that means that the module's internal memory can be exposed via those functions, so we'll still need to perform partial initialization.

This is actually also a slight bug I think with the pooling allocator which I think calls `set_instance_memories` first before `set_instance_tables`, but should be an easy fix to swap them.

I also feel like that while it's interesting-ish to think about this it's probably not really worth a complicated test to ensure the letter-of-the-law is followed in this respect.

Reply:
Isn't populating an imported table with functions from a module that failed to instantiate...problematic? The function might (transitively) reference a defined memory that wasn't initialized properly because the data segment never got copied because a previous segment was out of bounds:

> Each active segment is initialized in module-definition order. For each segment, if reading the source or writing the destination would go out of bounds, then instantiation fails at that point. Data that had already been written for previous (in-bounds) segments stays written.

I'm not exactly sure why the bulk memory proposal did away with the upfront bounds checking on the segments at initialization time and explicitly allowed for these side effects to occur; of which there is no reasonable recovery other than completely throwing out any table or memory that was imported to the module that failed to instantiate as they are now in indeterminate states.

Reply:
Re: the order of `set_instance_memories` and `set_instance_tables`, these are setting up the `Instance`'s fields and both done prior to initialization, so the order doesn't matter.

The order matters in `initialize_instance`, where `initialize_tables` is called before `initialize_memories` (with uffd enabled, `initialize_memories` isn't called so that the page fault handler can on-demand init the pages).

Reply:
IIRC the bulk memory change was to spec memory initialization as `memory.init` instructions to make the spec a bit simpler, but I won't pretend to know all the intricacies there. I do believe, though, that it was acknowledged that functions could be put into tables which could reference not-fully-initialized memories at the time.

For the ordering ah yeah good point, I think you're right in that a swap isn't necessary.

Reply:
So I am correcting this to not treat "out of bounds" and "paged" initialization as two different variants; the "paged" initialization now records if an out of bounds segment was observed in addition to the pages that, up until that point, were initialized.

The memory initialization process now first copies the "successfully initialized up until the out of bounds access" pages (or let's the uffd handler fill the pages as needed) before a trap is returned.  This should maintain the observable side-effects required by the bulk memory proposal.